### PR TITLE
Server-streaming support in gRPC server pseudo-Node

### PR DESCRIPTION
**Pull request by daviddrysdale**
# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [x] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs [hunk 1]
```diff
@@ -388,59 +399,65 @@ impl UnaryService<Vec<u8>> for GrpcRequestHandler {
...
-            debug!("Sending gRPC response: {:?}", response);
-            timer.observe_duration();
-            match response.status {
-                None => Ok(tonic::Response::new(response.rsp_msg)),
-                Some(status) if status.code == oak_abi::proto::google::rpc::Code::Ok as i32 => {
-                    Ok(tonic::Response::new(response.rsp_msg))
+            // First convert the `Iterator<Item = GrpcResponse>` to an
+            // `Iterator<Item = Result<Vec<u8>, tonic::Status>`.
+            let result_iter = response_iter.map(|response| {
+                debug!("Sending gRPC response: {:?}", response);
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs:419]: Is it sending a gRPC response here, or just remapping?

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs [hunk 2]
```diff
@@ -504,64 +521,126 @@ impl GrpcRequestHandler {
...
+        ))
+    }
+}
+
+struct GrpcResponseIterator {
+    runtime: RuntimeProxy,
+    response_reader: oak_abi::Handle,
+    metrics_data: Metrics,
+    method_name: String,
+    _timer: prometheus::HistogramTimer,
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs:539]: Why is it necessary?

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs [hunk 3]
```diff
@@ -504,64 +521,126 @@ impl GrpcRequestHandler {
...
+}
+
+impl GrpcResponseIterator {
+    fn new(
+        runtime: RuntimeProxy,
+        response_reader: oak_abi::Handle,
+        metrics_data: Metrics,
+        method_name: String,
+    ) -> Self {
+        debug!(
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs:550]: Is it better to make it `trace!` ?

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs [hunk 4]
```diff
@@ -504,64 +521,126 @@ impl GrpcRequestHandler {
...
         }
+    }
+}
 
-        Ok(response_reader)
+/// Manual implementation of the `Drop` trait to ensure the response channel
+/// is always closed.
+impl Drop for GrpcResponseIterator {
+    fn drop(&mut self) {
+        debug!(
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs:574]: It's a very low-level message, could also be `trace!`.

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs [hunk 5]
```diff
@@ -504,64 +521,126 @@ impl GrpcRequestHandler {
...
-                    })
-                })
-                .and_then(|response| {
-                    // Return the serialized message body.
-                    GrpcResponse::decode(response.as_slice()).map_err(|error| {
-                        error!("Couldn't parse the GrpcResponse message: {}", error);
-                    })
-                })
+                            .observe(grpc_rsp.rsp_msg.len() as f64);
+                        if grpc_rsp.last {
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs:610]: Will it be marked as `last` by the Oak Node that sends this stream?

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs (hunk 2 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs:541]: It records the time taken to process the responses; when the timer is dropped it records the duration. 

Added a comment.

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs (hunk 4 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs]: Done.

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs (hunk 3 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs]: Done.

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs (hunk 5 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs:614]: Yes if it's actually a [unary method](https://github.com/project-oak/oak/blob/master/sdk/rust/oak/src/grpc/mod.rs#L297-L300), but not necessarily if it's what turns out to be the last response for a server-streaming method.  Added a comment.

File: oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs (hunk 1 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs]: Reworded to "Returning".

**Pull request merged**